jl4-decision-service/src/Server.hs

```diff
@@ -0,0 +1,854 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Server (
+  -- * AppM
+  AppM,
+  DbState (..),
+  ValidatedFunction (..),
+
+  -- * Servant
+  OperationId,
+
+  -- * REST API
+  Api,
+  FunctionApi,
+  FunctionApi' (..),
+  SingleFunctionApi,
+  SingleFunctionApi' (..),
+  FunctionCrud,
+  FunctionCrud' (..),
+  handler,
+
+  -- * API json types
+  Parameters (..),
+  Parameter (..),
+  Function (..),
+  SimpleFunction (..),
+  SimpleResponse (..),
+  Reasoning (..),
+  ReasoningTree (..),
+  ResponseWithReason (..),
+  EvaluatorError (..),
+  FnLiteral (..),
+  EvalBackend (..),
+  FunctionImplementation (..),
+  FnArguments (..),
+  Outcomes (..),
+  OutcomeObject (..),
+  OutcomeStyle (..),
+  BatchRequest (..),
+  InputCase (..),
+  OutputCase (..),
+  BatchResponse (..),
+  OutputSummary (..),
+
+  -- * utilities
+  toDecl,
+) where
+
+import Backend.Api
+import Backend.Api as Api
+import qualified Backend.Jl4 as Jl4
+
+import qualified Chronos
+import Control.Applicative
+import Control.Concurrent.STM
+import Control.Exception (evaluate)
+import Control.Monad (forM, when)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Except
+import Control.Monad.Trans.Reader (ReaderT (..), asks)
+import Data.Aeson (FromJSON, FromJSONKey, ToJSON, ToJSONKey, (.:), (.=))
+import qualified Data.Aeson as Aeson
+import Data.Aeson.Combinators.Decode (Decoder)
+import qualified Data.Aeson.Combinators.Decode as ACD
+import qualified Data.Aeson.Key as Aeson
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString as BS
+import Data.Fixed
+import Data.Int
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Maybe as Maybe
+import Data.Scientific (Scientific)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Encoding as TL
+import qualified Data.Tuple.Extra as Tuple
+import Data.Typeable
+import qualified GHC.Clock as Clock
+import GHC.Generics
+import GHC.TypeLits
+import Servant
+import System.Timeout (timeout)
+
+-- ----------------------------------------------------------------------------
+-- Servant API
+-- ----------------------------------------------------------------------------
+
+data DbState = DbState
+  { functionDatabase :: TVar (Map Text ValidatedFunction)
```

Comment:
I think this could be an MVar.

Reply:
Could be, but why should it be? You mean less overhead?

Reply:
it's simpler if we don't actually need any of the transaction semantics.

Reply:
just a nit though, can just ignore it